keps/sig-auth/2799-reduction-of-secret-based-service-account-token/README.md

```diff
@@ -0,0 +1,463 @@
+
+### Goals
+
+- No auto-generation of secret-based service account token.
+- Removal of unused auto-generated secret-based service account tokens
+
+### Non-Goals
+
+- Removal of [explicitly requested secret-based service account tokens](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token).
+
+## Proposal
+
+- Change the service account control loop in Token Controller to not auto-create
+  secret for service accounts. At the same time, warn usage of auto-created
+  secret-based service account tokens and encourage users to use TokenRequest
+  API or manually-created secret-based service account tokens.
+- Purge unused auto-generated secret-based service account tokens.
+
+### User Stories (Optional)
+
+### Notes/Constraints/Caveats
+
+- A warning mechanism should be implemented to help users
+  migrate.
+- Auto generated secret-based service account tokens are those requested by
+  Token Controller.
+- Only clean up auto-generated tokens which:
+  - are not referenced by pods
+  - have not been used to authenticate for some duration (time duration or number of releases)
+- To consult active usage of secret-based tokens, metric
+  `serviceaccount_legacy_tokens_total` or audit annotation
+  `authentication.k8s.io/legacy-token` could be used.
+
+### Risks and Mitigations
+
+- When feature LegacyServiceAccountTokenNoAutoGeneration is Beta, consumers
+  depending directly on waiting for and reading tokens out of auto-generated
+  secrets might stop working. To mitigate,
+  1. Emit warnings when using auto-generated token secrets.
+  2. Publish pointers to TokenRequest or the manual secret request flow.
+- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
+  secret-based token might stop working. To mitigate,
+  1. When Alpha, annouce the cleanup starts at Beta
+  2. Emit warnings when using auto-generated token secrets.
+  3. Add pointers of TokenRequest API and manually created tokens in the validation
+     result.
+
+## Design Details
+
+### LegacyServiceAccountTokenNoAutoGeneration:
+
+Token Controller stops auto-creating secret for service accounts. This feature would
+be enabled when it is implemented since no new code is added and this can make
+sure new clusters are in good state.
+
+### LegacyServiceAccountTokenTracking
+
+To facilitate LegacyServiceAccountTokenCleanUp, we implement a simple controller
+in kube-apiserver that maintains a bool value configmap in `kube-system` to
+indicates if tracking is enabled in the cluster. It is similar to the existing
+`ClusterAuthenticationTrustController` that maintains `configmap/extension-apiserver-authentication`.
+
+- When LegacyServiceAccountTokenTracking is enabled in all apiservers,
+
+  - the controller creates/updates a configmap in `kube-system` namespace that
+    stores the current date as `tracked-since`.
+  - when a legacy token is used, issue a warning, annotate/update the
```

Comment:
you'll annotate legacy tokens on every usage?

Reply:
plan was to do so at date granularity... max of one write per 24 hour period

Reply:
captured in the KEP.